dev/ProgressBar/ProgressBar.cpp

```diff
@@ -0,0 +1,154 @@
+
+    SetValue(s_TemplateSettingsProperty, winrt::make<::ProgressBarTemplateSettings>());
+}
+
+void ProgressBar::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    // NOTE: Example of how named parts are loaded from the template. Important to remember that it's possible for
+    // any of them not to be found, since devs can replace the template with their own.
+    m_layoutRoot.set(GetTemplateChildT<winrt::Grid>(L"LayoutRoot", controlProtected));
+    if (auto layoutRoot = m_layoutRoot.get())
+    {
+        // NOTE: You can hook up to events or property changes on your template parts here, but you have to hook up revokers also
+        // because the template can change or the control could be removed and readded to the tree, etc.
+        // (I can think of no reason you would want to handle grid loading, this is just an example of how it works.)
+        m_layoutRootLoadedRevoker = layoutRoot.Loaded(winrt::auto_revoke, { this, &ProgressBar::OnLayoutRootLoaded });
+    }
+
+    m_progressBarIndicator.set(GetTemplateChildT<winrt::Rectangle>(L"ProgressBarIndicator", controlProtected));
+    if (auto progressBarIndicator = m_progressBarIndicator.get())
+    {
+        m_progressBarIndicatorRevoker = progressBarIndicator.Loaded(winrt::auto_revoke, { this, &ProgressBar::OnLayoutRootLoaded });
+    }
+
+    UpdateStates();
+}
+
+void ProgressBar::OnLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    // TODO: things
+}
+
+void ProgressBar::OnLayoutRootLoaded(const winrt::IInspectable&, const winrt::RoutedEventArgs&)
+{
+    // TODO: things
+}
+
+void ProgressBar::OnRangeBaseValueChanged(const winrt::DependencyObject& sender, const winrt::DependencyProperty& args)
+{
+    // NOTE: This hits when the Value property changes, because we called RegisterPropertyChangedCallback.
+
+    SetProgressBarIndicatorWidth();
+}
+
+void ProgressBar::OnIsIndeterminatePropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    // NOTE: This hits when IsIndeterminate changes because we set MUX_PROPERTY_CHANGED_CALLBACK to true in the idl.
+
+    UpdateStates();
+}
+
+void ProgressBar::OnShowPausedPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::OnShowErrorPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    UpdateStates();
+}
+
+void ProgressBar::UpdateStates()
+{
+    if (ShowError())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Error", true);
+    }
+    else if (ShowPaused() && IsIndeterminate())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Error", true); // Paused-Indeterminate state same visual treatment as Error state
+    }
+    else if (ShowPaused())
+    {
+        winrt::VisualStateManager::GoToState(*this, L"Paused", true);
+    }
+    else if (IsIndeterminate())
+    {
+        UpdateWidthBasedTemplateSettings();
```

Comment:
I don't think that you call this frequently enough.  If the control is put into intermediate mode we will set the width based template settings, but then if the width of the control changes we will not update the setting.  I think while we are in indeterminate mode we also need to be listening to the sizeChanged event and updating the template settings each time that fires.